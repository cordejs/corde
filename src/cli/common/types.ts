import { Command } from "commander";

export type ParamsFrom = "args" | "options" | "both";

export interface ICommandOptions {
  flags: string;
  description?: string;
  defaultValue?: string;
}

export interface ICliCommandConstructor {
  name?: string;
  program: Command;
  paramsFrom?: ParamsFrom;
}

export interface ICliCommand {
  readonly paramsFrom?: ParamsFrom;
  readonly command: Command;
  handler(...args: any[]): Promise<void> | void;
}

export interface ICliCommandConstructable {
  new (command: Command): ICliCommand;
}
