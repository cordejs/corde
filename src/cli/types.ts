export interface ICommand {
  action(...args: any[]): Promise<void> | void;
}
