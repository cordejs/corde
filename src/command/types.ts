import { ICordeBot } from "../types";

export interface ICreateMatcherParam {
  matcher?: string;
  isNot?: boolean;
  commandName?: string | boolean | number;
  channelId?: string;
  guildId?: string;
  isDebug?: boolean;
  isCascade?: boolean;
  cordeBot?: ICordeBot;
  trace?: string;
  utility?: boolean;
  mustSendCommand?: boolean;
}
