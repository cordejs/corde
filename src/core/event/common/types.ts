import { Client, ClientEvents } from "discord.js";

export type EventNames = keyof ClientEvents;

export type Spread<T extends EventNames> = ClientEvents[T][1] extends undefined
  ? ClientEvents[T] extends (infer U)[]
    ? U
    : never
  : ClientEvents[T];

export interface IDiscordEvent<TEvent extends EventNames, TFilter> {
  canListen(): boolean;
  getIntents(): string[];
  on(fn: (...args: ClientEvents[TEvent]) => void): this;
  once(filter?: TFilter): Promise<Spread<TEvent>>;
}

export interface IDiscordEventConstructable<TEvent extends EventNames, TFilter> {
  new (client: Client): IDiscordEvent<TEvent, TFilter>;
}

export type IntentType =
  | "GUILDS"
  | "GUILD_MEMBERS"
  | "GUILD_BANS"
  | "GUILD_EMOJIS_AND_STICKERS"
  | "GUILD_INTEGRATIONS"
  | "GUILD_WEBHOOKS"
  | "GUILD_INVITES"
  | "GUILD_VOICE_STATES"
  | "GUILD_PRESENCES"
  | "GUILD_MESSAGES"
  | "GUILD_MESSAGE_REACTIONS"
  | "GUILD_MESSAGE_TYPING"
  | "DIRECT_MESSAGES"
  | "DIRECT_MESSAGE_REACTIONS"
  | "DIRECT_MESSAGE_TYPING"
  | "GUILD_SCHEDULED_EVENTS";

export type EventClass = {
  [k in IntentType]: {
    bitfield: number;
    events: EventNames[];
  };
};
