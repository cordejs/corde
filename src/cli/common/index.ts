export * from "./loadConfigs";
export * from "./commandFactory";
