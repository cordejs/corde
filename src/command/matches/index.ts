export * from "./message";
export * from "./role";
