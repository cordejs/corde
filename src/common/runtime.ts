import { IConfigOptions, ICordeBot, TestFunctionType } from "../types";
import { Config } from "./config";
import { Client } from "discord.js";
import { CordeBot } from "../core/cordeBot";
import { ConfigError } from "../errors";

const Environment = {
  isUnityTest: process.env.ENV === "UNITY_TEST",
  isE2eTest: process.env.ENV === "E2E_TEST",
};

/**
 * @internal
 */
class Runtime implements IConfigOptions {
  get bot() {
    if (!this._bot) {
      this._bot = this.initBot();
    }

    return this._bot;
  }

  get isTestEnv() {
    return this.environment.isE2eTest || this.environment.isUnityTest;
  }

  get environment() {
    return Environment;
  }

  get exitOnFileReadingError() {
    return this._configs.exitOnFileReadingError;
  }

  get extensions() {
    return this._configs.extentions;
  }

  get events() {
    return this.bot.events;
  }

  get configs() {
    return this._configs;
  }

  get cordeBotToken() {
    return this._configs.cordeBotToken;
  }

  get botTestId() {
    return this._configs.botTestId;
  }

  get project() {
    return this._configs.project;
  }

  get botToken() {
    return this._configs.botToken;
  }

  get channelId() {
    return this._configs.channelId;
  }

  get guildId() {
    return this._configs.guildId;
  }

  get timeout() {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    return this._configs.timeout!;
  }

  get botPrefix() {
    return this._configs.botPrefix;
  }

  get testMatches() {
    return this._configs.testMatches;
  }

  get modulePathIgnorePatterns() {
    return this._configs.modulePathIgnorePatterns;
  }

  private constructor() {
    this._configs = new Config();
  }

  private static _instance: Runtime;
  configFilePath!: string;
  files!: string[];

  private readonly _configs: Config;
  private _bot!: ICordeBot;

  static getInstance() {
    if (!Runtime._instance) {
      Runtime._instance = new Runtime();
    }
    return Runtime._instance;
  }

  setConfigs(_configs: Partial<IConfigOptions>, forceUpdate?: boolean) {
    if (!_configs) {
      throw new ConfigError("Invalid _configs");
    }

    this._configs.setConfigs(_configs, forceUpdate);
  }

  /**
   * Shortcut for *bot.isLoggedIn*
   */
  isBotLoggedIn() {
    return this.bot && this.bot.isLoggedIn();
  }

  /**
   * Shortcut for *bot.logout*
   */
  logoffBot() {
    if (this._bot) {
      this._bot.logout();
    }
  }

  async loginBot(token: string) {
    return await this.bot.login(token);
  }

  injectBot(fn: TestFunctionType) {
    return fn(this.bot);
  }

  initBot() {
    return new CordeBot(
      this._configs.botPrefix,
      this._configs.guildId,
      this._configs.channelId,
      this._configs.botTestId,
      new Client(),
    );
  }
}

/**
 * Singleton of Runtime.
 */
const runtime = Runtime.getInstance();
export { runtime };
