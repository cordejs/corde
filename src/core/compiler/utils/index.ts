export * from "./import-js-in-directory";
export * from "./was-file-modified";
