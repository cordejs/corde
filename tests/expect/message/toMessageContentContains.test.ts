import { Client } from "discord.js";
import MockDiscord from "../../mocks/mockDiscord";
import { createCordeBotWithMockedFunctions, testUtils } from "../../testHelper";
import { MockEvents } from "../../mocks/mockEvents";
import { ICordeBot, ITestReport } from "../../../src/types";
import { ToMessageContentContains } from "../../../src/expect/matches";
import { buildReportMessage } from "../../../src/utils";
import { runtime } from "../../../src/common/runtime";

let mockDiscord = new MockDiscord();

describe("testing toMessageContentContains", () => {
  afterEach(() => {
    mockDiscord = new MockDiscord();
  });

  function initTestClass(cordeBot: ICordeBot, isNot: boolean) {
    return testUtils.initTestClass(ToMessageContentContains, {
      isCascade: false,
      command: "toMessageContentContains",
      cordeBot: cordeBot,
      isNot: isNot,
    });
  }

  it("should return a failed test due to invalid parameter (null)", async () => {
    const cordeClient = createCordeBotWithMockedFunctions(mockDiscord, new Client());

    const message = buildReportMessage(`expected content can not be null or empty`);

    const toMessageContentContains = initTestClass(cordeClient, false);

    const reportModel: ITestReport = {
      pass: false,
      testName: toMessageContentContains.toString(),
      message,
    };

    const report = await toMessageContentContains.action(null);
    expect(report).toEqual(reportModel);
    expect(report).toMatchSnapshot();
  });

  it("should return a failed test due to invalid parameter (undefined)", async () => {
    const cordeClient = createCordeBotWithMockedFunctions(mockDiscord, new Client());
    const message = buildReportMessage(`expected content can not be null or empty`);
    const toMessageContentContains = initTestClass(cordeClient, false);

    const reportModel: ITestReport = {
      pass: false,
      testName: toMessageContentContains.toString(),
      message,
    };

    const report = await toMessageContentContains.action(undefined);
    expect(report).toEqual(reportModel);
    expect(report).toMatchSnapshot();
  });

  it("should fail due to no message was sent by the bot", async () => {
    runtime.setConfigs({ timeOut: 100 }, true);
    const cordeClient = createCordeBotWithMockedFunctions(mockDiscord, new Client());

    const message = buildReportMessage(
      `expected: testing bot to send a message\n`,
      `received: no message was sent`,
    );

    const toMessageContentContains = initTestClass(cordeClient, false);

    const reportModel: ITestReport = {
      pass: false,
      testName: toMessageContentContains.toString(),
      message,
    };

    const report = await toMessageContentContains.action("pong");
    expect(report).toEqual(reportModel);
    expect(report).toMatchSnapshot();
  });

  it("should get success test due to timeout but isNot = true", async () => {
    runtime.setConfigs({ timeOut: 100 }, true);
    const cordeClient = createCordeBotWithMockedFunctions(mockDiscord, new Client());

    const toMessageContentContains = initTestClass(cordeClient, true);

    const reportModel: ITestReport = {
      pass: true,
      testName: toMessageContentContains.toString(),
    };

    const report = await toMessageContentContains.action("pong");
    expect(report).toEqual(reportModel);
    expect(report).toMatchSnapshot();
  });

  it("should get success test due to bot returned equal message", async () => {
    runtime.setConfigs({ timeOut: 100 }, true);
    const cordeClient = createCordeBotWithMockedFunctions(mockDiscord, new Client());

    const events = new MockEvents(cordeClient, mockDiscord);
    events.mockOnceMessage();

    const toMessageContentContains = initTestClass(cordeClient, false);

    const reportModel: ITestReport = {
      pass: true,
      testName: toMessageContentContains.toString(),
    };

    const report = await toMessageContentContains.action(mockDiscord.message.content);

    expect(report).toEqual(reportModel);
    expect(report).toMatchSnapshot();
  });

  it("should get success test due to bot returned a contained message", async () => {
    runtime.setConfigs({ timeOut: 100 }, true);
    const cordeClient = createCordeBotWithMockedFunctions(mockDiscord, new Client());

    const events = new MockEvents(cordeClient, mockDiscord);
    mockDiscord.message.content = "message to contain";
    events.mockOnceMessage();

    const toMessageContentContains = initTestClass(cordeClient, false);

    const reportModel: ITestReport = {
      pass: true,
      testName: toMessageContentContains.toString(),
    };

    const report = await toMessageContentContains.action("message");

    expect(report).toEqual(reportModel);
    expect(report).toMatchSnapshot();
  });

  it("should get fail test due to bot returned different messages both primitive values", async () => {
    runtime.setConfigs({ timeOut: 100 }, true);
    const cordeClient = createCordeBotWithMockedFunctions(mockDiscord, new Client());

    const events = new MockEvents(cordeClient, mockDiscord);
    events.mockOnceMessage();

    const toMessageContentContains = initTestClass(cordeClient, false);
    const expectValue = "expect value";

    const reportModel: ITestReport = {
      pass: false,
      testName: toMessageContentContains.toString(),
      message: buildReportMessage(
        `expected '${mockDiscord.message.content}' to contains '${expectValue}'`,
      ),
    };

    const report = await toMessageContentContains.action(expectValue);

    expect(report).toEqual(reportModel);
    expect(report).toMatchSnapshot();
  });

  it("should return a failed test due to failure in message sending", async () => {
    const corde = createCordeBotWithMockedFunctions(mockDiscord, new Client());
    corde.getRoles = jest.fn().mockReturnValue(mockDiscord.roleManager.cache);
    corde.findRole = jest.fn().mockReturnValue(mockDiscord.role);

    const erroMessage = "can not send message to channel x";
    corde.sendTextMessage = jest
      .fn()
      .mockImplementation(() => Promise.reject(new Error(erroMessage)));

    const toMessageContentContains = initTestClass(corde, false);
    const report = await toMessageContentContains.action("abc");

    const reportModel: ITestReport = {
      pass: false,
      message: buildReportMessage(erroMessage),
      testName: toMessageContentContains.toString(),
    };

    expect(report).toEqual(reportModel);
    expect(report).toMatchSnapshot();
  });
});
