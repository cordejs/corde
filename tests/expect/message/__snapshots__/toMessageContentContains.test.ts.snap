// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing toMessageContentContains should fail due to no message was sent by the bot 1`] = `
Object {
  "message": "    
    expected: testing bot to send a message
    received: no message was sent
",
  "pass": false,
  "testName": "toMessageContentContains",
}
`;

exports[`testing toMessageContentContains should get fail test due to bot returned different messages both primitive values 1`] = `
Object {
  "message": "    
    expected 'this is the message content' to contains 'expect value'
",
  "pass": false,
  "testName": "toMessageContentContains",
}
`;

exports[`testing toMessageContentContains should get success test due to bot returned a contained message 1`] = `
Object {
  "pass": true,
  "testName": "toMessageContentContains",
}
`;

exports[`testing toMessageContentContains should get success test due to bot returned equal message 1`] = `
Object {
  "pass": true,
  "testName": "toMessageContentContains",
}
`;

exports[`testing toMessageContentContains should get success test due to timeout but isNot = true 1`] = `
Object {
  "pass": true,
  "testName": "toMessageContentContains",
}
`;

exports[`testing toMessageContentContains should return a failed test due to failure in message sending 1`] = `
Object {
  "message": "    
    can not send message to channel x
",
  "pass": false,
  "testName": "toMessageContentContains",
}
`;

exports[`testing toMessageContentContains should return a failed test due to invalid parameter (null) 1`] = `
Object {
  "message": "    
    expected content can not be null or empty
",
  "pass": false,
  "testName": "toMessageContentContains",
}
`;

exports[`testing toMessageContentContains should return a failed test due to invalid parameter (undefined) 1`] = `
Object {
  "message": "    
    expected content can not be null or empty
",
  "pass": false,
  "testName": "toMessageContentContains",
}
`;
